lib/builtin.t

```diff
@@ -94,6 +94,62 @@ package FetchStoreCounter {
     is(blessed(bless [], "0") ? "YES" : "NO", "NO", 'blessed in boolean context handles "0" cornercase');
 }
 
+# created_as_...
+{
+    use builtin qw( created_as_string created_as_number );
+
+    # some literal constants
+    ok(!created_as_string(undef), 'undef created as !string');
+    ok(!created_as_number(undef), 'undef created as !number');
+
+    ok( created_as_string("abc"), 'abc created as string');
+    ok(!created_as_number("abc"), 'abc created as number');
+
+    ok(!created_as_string(123),   '123 created as !string');
+    ok( created_as_number(123),   '123 created as !number');
+
+    ok(!created_as_string(1.23),   '1.23 created as !string');
+    ok( created_as_number(1.23),   '1.23 created as !number');
+
+    ok(!created_as_string([]),    '[] created as !string');
+    ok(!created_as_number([]),    '[] created as !number');
+
+    ok(!created_as_string(builtin::true), 'true created as !string');
+    ok(!created_as_number(builtin::true), 'true created as !number');
+
+    ok(builtin::is_bool(created_as_string(0)), 'created_as_string returns bool');
+    ok(builtin::is_bool(created_as_number(0)), 'created_as_number returns bool');
+
+    # variables
+    my $just_str = "def";
+    ok( created_as_string($just_str), 'def created as string');
+    ok(!created_as_number($just_str), 'def created as number');
+
+    my $just_num = 456;
+    ok(!created_as_string($just_num), '456 created as string');
+    ok( created_as_number($just_num), '456 created as number');
+
+    # variables reused
+    my $originally_str = "1";
+    my $str_as_num = $originally_str + 0;
+    ok( created_as_string($originally_str), 'PV reused as IV created as string');
+    ok(!created_as_number($originally_str), 'PV reused as IV created as !number');
+    ok(!created_as_string($str_as_num), 'New number from PV created as !string');
+    ok( created_as_number($str_as_num), 'New number from PV created as number');
+
+    my $originally_num = 1;
+    my $num_as_str = "$originally_num";
+    ok(!created_as_string($originally_num), 'IV reused as PV created as !string');
+    ok( created_as_number($originally_num), 'IV reused as PV created as number');
+    ok( created_as_string($num_as_str), 'New string from IV created as string');
+    ok(!created_as_number($num_as_str), 'New string from IV created as !number');
+
+    # magic
+    local $1;
+    "hello" =~ m/(.*)/;
+    ok(created_as_string($1), 'magic string');
+}
+
```

Comment:
How do we expect this to behave? Right now it's somewhat surprising:
```
$ cat ex.pl 
#!./perl -I lib

use strict;
use warnings;

use builtin qw(created_as_number created_as_string);
no warnings qw(experimental);

my $x = 2**10000000;

if (created_as_number($x)) {
  print "created as number\n";
}

if (created_as_string($x)) {
  print "created as string\n";
}

my $z = "$x";

if (created_as_number($x)) {
  print "created as number\n";
}

if (created_as_string($x)) {
  print "created as string\n";
}
```
```
created as number
created as string
```

Here's one case where using the value as a string modifies it behind the scenes

Reply:
This seems like a bug. @nwc10 could you take a look at this? It looks like the branches in  `Perl_sv_2pv_flags` for Inf/NaN or when USE_LOCALE_NUMERIC is not defined are still setting SvPOK_on.

Reply:
Ooooh oops, that looks like it might be a bug :/ That's supposed to remain numeric

Reply:
Turns out that was indeed a bug in core perl; #19540 fixed it. This PR has been rebased on that and I've added some more tests for those cases.